bin/src/lib.rs

```diff
@@ -292,12 +299,69 @@ where
 
 /// Configuration for a zone
 #[derive(Deserialize, Debug)]
-#[serde(deny_unknown_fields)]
 pub struct ZoneConfig {
     /// name of the zone
     pub zone: String, // TODO: make Domain::Name decodable
     /// type of the zone
-    pub zone_type: ZoneType,
+    #[serde(flatten)]
+    pub zone_type_config: ZoneTypeConfig,
+}
+
+impl ZoneConfig {
+    /// Return a new zone configuration
+    ///
+    /// # Arguments
+    ///
+    /// * `zone` - name of a zone, e.g. example.com
+    /// * `zone_type_configuration` - Configuration specific to the type of zone, e.g. Primary, Secondary, etc.
+    pub fn new(zone: String, zone_type_config: ZoneTypeConfig) -> Self {
+        Self {
+            zone,
+            zone_type_config,
+        }
+    }
+
+    // TODO this is a little ugly for the parse, b/c there is no terminal char
+    /// returns the name of the Zone, i.e. the `example.com` of `www.example.com.`
+    pub fn zone(&self) -> Result<Name, ProtoError> {
+        Name::parse(&self.zone, Some(&Name::new()))
+    }
+
+    /// the type of the zone
+    pub fn zone_type(&self) -> ZoneType {
+        match &self.zone_type_config {
+            ZoneTypeConfig::Primary { .. } => ZoneType::Primary,
+            ZoneTypeConfig::Secondary { .. } => ZoneType::Secondary,
+            ZoneTypeConfig::Forward { .. } => ZoneType::Forward,
+            ZoneTypeConfig::Hint { .. } => ZoneType::Hint,
+        }
+    }
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(tag = "zone_type")]
+#[serde(deny_unknown_fields)]
+/// Enumeration over each zone type's configuration.
+pub enum ZoneTypeConfig {
+    Primary(ServerZoneConfig),
+    Secondary(ServerZoneConfig),
+    Forward(ForwardZoneConfig),
+    Hint(HintZoneConfig),
```

Comment:
Could these be named field variants instead so we could skip the intermediate type definitions?